packages/ui/src/elements/Table/index.tsx

```diff
@@ -5,55 +5,155 @@ import type { Column } from 'payload'
 import React from 'react'
 
 import './index.scss'
+import { useListQuery } from '../../providers/ListQuery/index.js'
+import { DraggableSortableItem } from '../DraggableSortable/DraggableSortableItem/index.js'
+import { DraggableSortable } from '../DraggableSortable/index.js'
 
 const baseClass = 'table'
 
+// if you change this, you need to change it in a couple of places where it's duplicated
+const ORDER_FIELD_NAME = 'payload-order'
+
 export type Props = {
   readonly appearance?: 'condensed' | 'default'
   readonly columns?: Column[]
-  readonly data: Record<string, unknown>[]
+  readonly data: { [key: string]: unknown; id: string; 'payload-order': string }[]
 }
 
-export const Table: React.FC<Props> = ({ appearance, columns, data }) => {
+export const Table: React.FC<Props> = ({ appearance, columns, data: initialData }) => {
+  const { handleSortChange, query } = useListQuery()
+  const [data, setData] = React.useState(initialData)
+
+  // Force re-sort when data changes
+  React.useEffect(() => {
+    if (query.sort) {
+      void handleSortChange(query.sort as string).catch((error) => {
+        throw error
+      })
+    }
+  }, [data, handleSortChange, query.sort])
+
   const activeColumns = columns?.filter((col) => col?.active)
 
   if (!activeColumns || activeColumns.length === 0) {
     return <div>No columns selected</div>
   }
 
+  const handleDragEnd = async ({ moveFromIndex, moveToIndex }) => {
+    if (moveFromIndex === moveToIndex) {
+      return
+    }
+
+    const movedId = data[moveFromIndex].id
+    const newBeforeRow = moveToIndex > moveFromIndex ? data[moveToIndex] : data[moveToIndex - 1]
+    const newAfterRow = moveToIndex > moveFromIndex ? data[moveToIndex + 1] : data[moveToIndex]
+
+    // To debug:
+    // console.log(
+    //   `moving ${data[moveFromIndex]?.text} between ${newBeforeRow?.text} and ${newAfterRow?.text}`,
+    // )
+
+    // Store the original data for rollback
+    const previousData = [...data]
+
+    // TODO: this optimistic update is not working (the table is not re-rendered)
+    // you can't debug it commenting the try block. Every move needs to be followed by
+    // a refresh of the page to see the changes.
+    setData((currentData) => {
+      const newData = [...currentData]
+      newData[moveFromIndex] = {
+        ...newData[moveFromIndex],
+        [ORDER_FIELD_NAME]: `${newBeforeRow?.[ORDER_FIELD_NAME]}_pending`,
+      }
+      // move from index to moveToIndex
+      newData.splice(moveToIndex, 0, newData.splice(moveFromIndex, 1)[0])
+
+      return newData
+    })
+
+    try {
+      // Assuming we're in the context of a collection
+      const collectionSlug = window.location.pathname.split('/').filter(Boolean)[2]
```

Comment:
This will not be reliable, we will have to come up with something different. Maybe Table can take a handleDrag prop? We could export a function that creates this function so it can be easily re-used. 

I think adding a onDragEnd prop would make this component more composible as well and not tied to payload specifics.

Reply:
I tried this very hard, but it was not possible. dragEnd is very rooted with a lot of state available only in that component. Also, it is called by `renderTable` which is not a component, but a function that runs on the server. There you cannot make hooks or implement for example optimistic refresh.
Anyway, the Table component is already very tied to Payload specifics, so I don't think that is problem.

Reply:
If I have basePath or change my admin route to `/nested/admin` this will break.